pkg/upgrade/upgrades/update_invalid_column_ids_in_sequence_back_references.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package upgrades
+
+import (
+	"context"
+	"reflect"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
+	"github.com/cockroachdb/cockroach/pkg/upgrade"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+)
+
+func updateInvalidColumnIDsInSequenceBackReferences(
+	ctx context.Context, _ clusterversion.ClusterVersion, d upgrade.TenantDeps, _ *jobs.Job,
+) error {
+	// Repeatedly start a new transaction that scans all sequences and update invalid
+	// column IDs, if any, in their back references, until there are no repairs needed.
+	for {
+		hasUpgrade := false
+
+		if err := d.CollectionFactory.TxnWithExecutor(ctx, d.DB, d.SessionData, func(
+			ctx context.Context, txn *kv.Txn, descriptors *descs.Collection, ie sqlutil.InternalExecutor,
+		) error {
+			// Upgrade each sequence, one at a time, until we exhaust all of them.
+			var lastUpgradedID descpb.ID
```

Comment:
@ajwerner I've changed how we paginate the descriptors during repair and now battling CI flakes, but think you might want to take one last look again, just in case.